src/test/ui/cmse-nonsecure-call/params-on-registers.rs

```diff
@@ -0,0 +1,15 @@
+// build-pass
+// compile-flags: --target thumbv8m.main-none-eabi --crate-type lib
+// only-thumbv8m.main-none-eabi
+#![feature(abi_cmse_nonsecure_call)]
+#![no_std]
+
+#[no_mangle]
+pub fn test(a: u32, b: u32, c: u32, d: u32) -> u32 {
+    let non_secure_function = unsafe {
+        core::mem::transmute::<usize, extern "cmse-nonsecure-call" fn(u32, u32, u32, u32) -> u32>(
```

Comment:
Can you also add a test that defines a `extern "cmse-nonsecure-call" fn` in Rust instead of transmuting?

Reply:
I tried the following (if that is what you meant):
```rust
#![no_std]
#![feature(abi_cmse_nonsecure_call)]

#[no_mangle]
pub fn test() -> u32 {
    toto()
}

extern "cmse-nonsecure-call" fn toto() -> u32 {
    45
}
```
and I have the following LLVM error:
```
[hugdev01@machine tests]$ rustc +stage1 --target thumbv8m.main-none-eabi --crate-type lib --emit asm toto.rs
rustc: /home/hugdev01/dev/rust-work/rust/src/llvm-project/llvm/lib/Target/ARM/ARMISelLowering.cpp:2556: virtual llvm::SDValue llvm::ARMTargetLowering::LowerCall(llvm::TargetLowering::CallLoweringInfo&, llvm::SmallVectorImpl<llvm::SDValue>&) const: Assertion `!isARMFunc && !isDirect && "Cannot handle call to ARM function or direct call"' failed.
Aborted (core dumped)
```
I believe this is because it is forbidden, in LLVM this callsite attribute can really only be used on function pointers but not directly on function declaration.
The [specification document](https://developer.arm.com/documentation/ecm0359818/latest/) says that section 5.5:
> A non-secure function type must only be used as a base type of a pointer. This disallows function definitions with this attribute and ensures a secure executable file only contains secure function definitions.

I guess ideally this kind of error should be returned by Rust instead of LLVM.

Reply:
Ah, I see. So the intention is that secure and non-secure worlds are always entirely separate executables, and `cmse-nonsecure-call` is used by the secure side only to invoke callback functions passed by the non-secure side?

Where can I find documentation of the LLVM implementation? https://llvm.org/docs/LangRef.html doesn't mention it.

(I agree that this should be checked by rustc instead of LLVM, especially since this is not a diagnostic but an assertion failure)

Reply:
> So the intention is that secure and non-secure worlds are always entirely separate executables

Yes true!

> cmse-nonsecure-call is used by the secure side only to invoke callback functions passed by the non-secure side?

Yes I believe that is the main use-case! It's also possible for the Secure side to call functions that it knows are at particular addresses, for example the non-secure reset handler.

> Where can I find documentation of the LLVM implementation? https://llvm.org/docs/LangRef.html doesn't mention it.

There isn't much documentation of how those features are implemented in LLVM 😅. The changes were introduced by this [mail](https://lists.llvm.org/pipermail/cfe-dev/2019-March/061834.html) on the mailing list a long time ago. Apart from that there are the [specifications](https://developer.arm.com/documentation/ecm0359818/latest) and the various commits on LLVM to get an idea of how it's implemented. Not ideal, I know.

> (I agree that this should be checked by rustc instead of LLVM, especially since this is not a diagnostic but an assertion failure)

I am going to look where it is possible to check the position of the `extern` keywork to differentiate between function declaration/definition and function pointer type.

Reply:
I found a way to do this in the AST visitor it seems, by checking if the ABI is applied on a bare function or not. Might be a bit hacky to modify function prototype just for this feature. But it works!